### DISCUSSION TITLE
Non-null child field resolution

### DISCUSSION THREAD

**Discussion by dan-f** [Netflix/dgs-framework#1287] [Q&A]
Hi,

I have what I hope is a simple question. Consider the following schema:

```graphql
type Parent {
  id: ID!
  child: Child!
}

type Child {
  id: ID!
}
```

...and the following kotlin classes, perhaps code-generated from the schema:

```kt
data class Parent(
  val id: String,
  val child: Child,
)

data class Child(
  val id: String,
)
```

Assume it's expensive to compute the child. We implement it via a separate resolver, making sure that it won't be invoked unless requested. That would look something like this:

```kt
@DgsComponent
class ParentDataFetcher(
  val childService: ChildService,
  val parentService: ParentService,
) {
  @DgsQuery
  fun parent(
    dfe: DgsDataFetchingEnvironment,
    @InputArgument id: String
  ): CompletableFuture<Recipient> {
    val internalParent = parentService.getById(id)
    return Parent(id = internalParent.id) // type error - need to supply child
  }

  @DgsData(parentType = "Parent", field = "child")
  fun child(dfe: DgsDataFetchingEnvironment) {
    val internalChild = childService.expensivelyLoadChildByParentId(dfe.getSource().id)
    return Child(id = internalChild.id)
  }
}
```

Kotlin rightly rejects this -- we can't construct `Parent` without providing a `Child` in the `parent` resolver. The obvious "fix" is to make the child nullable in our schema, but that doesn't actually correctly reflect the domain model in this case.

So what's the recommended approach? Should we be constructing trivial `Child`s on the parent's field resolver, perhaps mocking them? For lists, it's easy enough to provide an empty list, but most non-trivial objects will require data to construct. This feels like a common situation. Would appreciate any pointers!

**Comment by srinivasankavitha** [comment]
If the child field is optional, you can represent it by making the child field optional in the schema:
```
type Parent {
  id: ID!
  child: Child
}

type Child {
  id: ID!
}
```
Then, if you are using codegen, you can specify `generateKotlinNullableClasses = true` option to ensure it generates:
```
data class Parent(
  val id: String,
  val child?: Child
)

data class Child(
  val id: String
)
```

Hope that helps.

**Reply by dan-f** [reply]
The child field is _not_ optional, as mentioned in the original post :-)

**Reply by srinivasankavitha** [reply]
You can still use `generateKotlinNullableClasses` with codegen and that should generate the POJOs with optional fields.
